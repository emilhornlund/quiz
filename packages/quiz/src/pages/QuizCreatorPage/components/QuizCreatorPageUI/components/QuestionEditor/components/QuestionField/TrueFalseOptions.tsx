import React, { FC, useCallback, useEffect, useRef, useState } from 'react'

import { TextField } from '../../../../../../../../components'

import styles from './QuestionField.module.scss'

export interface TrueFalseOptionsProps {
  value?: boolean
  onChange: (value?: boolean) => void
  onValid: (valid: boolean) => void
}

const TrueFalseOptions: FC<TrueFalseOptionsProps> = ({
  value,
  onChange,
  onValid,
}) => {
  const [options, setOptions] = useState<
    { id: 'true' | 'false'; value: string; correct: boolean }[]
  >(() => [
    { id: 'true', value: 'True', correct: value === true },
    { id: 'false', value: 'False', correct: value === false },
  ])

  useEffect(() => {
    setOptions((prev) =>
      prev.map((opt) => ({
        ...opt,
        correct: opt.id === 'true' ? value === true : value === false,
      })),
    )
  }, [value])

  const handleChange = useCallback(
    (updatedIndex: number, checked: boolean) => {
      setOptions((prev) => {
        const next = prev.map((opt, i) =>
          i === updatedIndex
            ? { ...opt, correct: checked }
            : { ...opt, correct: checked ? false : opt.correct },
        )

        const picked = next.find((o) => o.correct)
        onChange(picked ? picked.id === 'true' : undefined)
        return next
      })
    },
    [onChange],
  )

  const handleAdditionalValidation = useCallback((): boolean | string => {
    const someCorrect = options.some((o) => o.correct)
    if (!someCorrect) return 'At least one option must be marked correct'
    return true
  }, [options])

  const [validById, setValidById] = useState<Record<string, boolean>>({})

  const handleValidChange = useCallback((id: string, valid: boolean) => {
    setValidById((prev) =>
      prev[id] === valid ? prev : { ...prev, [id]: valid },
    )
  }, [])

  const wasAllValid = useRef<boolean | undefined>(undefined)
  useEffect(() => {
    const isAllValid = options.every((o) => !!validById[o.id])
    if (wasAllValid.current !== isAllValid) {
      wasAllValid.current = isAllValid
      onValid(isAllValid)
    }
  }, [options, validById, onValid])

  // optional housekeeping if you ever change options shape dynamically:
  useEffect(() => {
    const allowed = new Set(options.map((o) => o.id))
    setValidById((prev) => {
      let changed = false
      const next: Record<string, boolean> = {}
      for (const k of Object.keys(prev)) {
        if (allowed.has(k as 'true' | 'false')) next[k] = prev[k]
        else changed = true
      }
      return changed ? next : prev
    })
  }, [options])

  return (
    <div className={styles.optionsContainer}>
      {options.map((option, index) => (
        <div key={`true-false-option-${index}`} className={styles.option}>
          <div className={styles.content}>
            <TextField
              id={`true-false-option-${index}-textfield`}
              type="text"
              value={option.value}
              checked={option.correct}
              onCheck={(checked) => handleChange(index, checked)}
              onValid={(ok) => handleValidChange(option.id, ok)}
              onAdditionalValidation={() => handleAdditionalValidation()}
              forceValidate
              readOnly
            />
          </div>
        </div>
      ))}
    </div>
  )
}

export default TrueFalseOptions
