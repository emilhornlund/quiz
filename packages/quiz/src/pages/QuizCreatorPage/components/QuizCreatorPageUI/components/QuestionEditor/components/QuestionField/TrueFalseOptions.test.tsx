import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'

import TrueFalseOptions from './TrueFalseOptions'

vi.mock('../../../../../../../../components', () => {
  return {
    TextField: (props: {
      id: string
      value: string
      checked?: boolean
      onCheck?: (c: boolean) => void
      onValid?: (ok: boolean) => void
      onAdditionalValidation?: () => boolean | string
      forceValidate?: boolean
      type?: string
      readOnly?: boolean
    }) => {
      const { id, checked, onCheck, onValid, onAdditionalValidation } = props

      const runValidation = () => {
        if (!onValid) return
        const a = onAdditionalValidation ? onAdditionalValidation() : true
        onValid(a === true)
      }

      const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        onCheck?.(e.target.checked)
      }

      React.useEffect(() => {
        runValidation()
        // revalidate when selection or validation rule changes
      }, [checked, onAdditionalValidation])

      return (
        <div data-testid={`tf-${id}`}>
          <input
            data-testid={`check-${id}`}
            type="checkbox"
            checked={!!checked}
            onChange={handleChange}
          />
        </div>
      )
    },
  }
})

function lastCallArg<T>(
  mock: { calls: unknown[][] },
  argIndex = 0,
): T | undefined {
  const calls = mock.calls
  if (!calls.length) return undefined

  return calls[calls.length - 1][argIndex] as T
}

describe('TrueFalseOptions', () => {
  const user = userEvent.setup()

  beforeEach(() => {
    vi.clearAllMocks()
  })

  afterEach(() => {
    vi.clearAllMocks()
  })

  it('renders and matches snapshot', () => {
    const onChange = vi.fn()
    const onValid = vi.fn()
    const { container } = render(
      <TrueFalseOptions onChange={onChange} onValid={onValid} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('initial validity is false when nothing selected', async () => {
    const onChange = vi.fn()
    const onValid = vi.fn()
    render(<TrueFalseOptions onChange={onChange} onValid={onValid} />)

    await waitFor(() => {
      const v = lastCallArg<boolean>(onValid.mock)
      expect(v).toBe(false)
    })
  })

  it('selecting True emits onChange(true) and sets valid=true', async () => {
    const onChange = vi.fn()
    const onValid = vi.fn()
    render(<TrueFalseOptions onChange={onChange} onValid={onValid} />)

    const trueBox = screen.getByTestId('check-true-false-option-0-textfield')
    await user.click(trueBox)

    const lastChange = lastCallArg<boolean | undefined>(onChange.mock)
    expect(lastChange).toBe(true)

    await waitFor(() => {
      const v = lastCallArg<boolean>(onValid.mock)
      expect(v).toBe(true)
    })
  })

  it('selecting False after True emits onChange(false) and keeps valid=true', async () => {
    const onChange = vi.fn()
    const onValid = vi.fn()
    render(<TrueFalseOptions onChange={onChange} onValid={onValid} />)

    const trueBox = screen.getByTestId('check-true-false-option-0-textfield')
    const falseBox = screen.getByTestId('check-true-false-option-1-textfield')

    await user.click(trueBox)
    await user.click(falseBox)

    const lastChange = lastCallArg<boolean | undefined>(onChange.mock)
    expect(lastChange).toBe(false)

    await waitFor(() => {
      const v = lastCallArg<boolean>(onValid.mock)
      expect(v).toBe(true)
    })
  })

  it('unchecking the active selection emits onChange(undefined) and sets valid=false', async () => {
    const onChange = vi.fn()
    const onValid = vi.fn()
    render(<TrueFalseOptions onChange={onChange} onValid={onValid} />)

    const trueBox = screen.getByTestId('check-true-false-option-0-textfield')

    await user.click(trueBox)
    await user.click(trueBox)

    const lastChange = lastCallArg<boolean | undefined>(onChange.mock)
    expect(lastChange).toBeUndefined()

    await waitFor(() => {
      const v = lastCallArg<boolean>(onValid.mock)
      expect(v).toBe(false)
    })
  })

  it('reflects external value changes (rerender) and stays valid=true when value provided', async () => {
    const onChange = vi.fn()
    const onValid = vi.fn()
    const { rerender, container } = render(
      <TrueFalseOptions onChange={onChange} onValid={onValid} />,
    )
    expect(container).toMatchSnapshot()

    rerender(
      <TrueFalseOptions value={true} onChange={onChange} onValid={onValid} />,
    )

    await waitFor(() => {
      const v = lastCallArg<boolean>(onValid.mock)
      expect(v).toBe(true)
    })
    expect(container).toMatchSnapshot()

    rerender(
      <TrueFalseOptions value={false} onChange={onChange} onValid={onValid} />,
    )

    await waitFor(() => {
      const v = lastCallArg<boolean>(onValid.mock)
      expect(v).toBe(true)
    })
    expect(container).toMatchSnapshot()
  })
})
