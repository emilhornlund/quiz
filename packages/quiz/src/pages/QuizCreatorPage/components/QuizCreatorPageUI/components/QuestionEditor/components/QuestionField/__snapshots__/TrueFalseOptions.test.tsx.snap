// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TrueFalseOptions > reflects external value changes (rerender) and stays valid=true when value provided 1`] = `
<div>
  <div
    class="optionsContainer"
  >
    <div
      class="option"
    >
      <div
        class="content"
      >
        <div
          data-testid="tf-true-false-option-0-textfield"
        >
          <input
            data-testid="check-true-false-option-0-textfield"
            type="checkbox"
          />
        </div>
      </div>
    </div>
    <div
      class="option"
    >
      <div
        class="content"
      >
        <div
          data-testid="tf-true-false-option-1-textfield"
        >
          <input
            data-testid="check-true-false-option-1-textfield"
            type="checkbox"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TrueFalseOptions > reflects external value changes (rerender) and stays valid=true when value provided 2`] = `
<div>
  <div
    class="optionsContainer"
  >
    <div
      class="option"
    >
      <div
        class="content"
      >
        <div
          data-testid="tf-true-false-option-0-textfield"
        >
          <input
            data-testid="check-true-false-option-0-textfield"
            type="checkbox"
          />
        </div>
      </div>
    </div>
    <div
      class="option"
    >
      <div
        class="content"
      >
        <div
          data-testid="tf-true-false-option-1-textfield"
        >
          <input
            data-testid="check-true-false-option-1-textfield"
            type="checkbox"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TrueFalseOptions > reflects external value changes (rerender) and stays valid=true when value provided 3`] = `
<div>
  <div
    class="optionsContainer"
  >
    <div
      class="option"
    >
      <div
        class="content"
      >
        <div
          data-testid="tf-true-false-option-0-textfield"
        >
          <input
            data-testid="check-true-false-option-0-textfield"
            type="checkbox"
          />
        </div>
      </div>
    </div>
    <div
      class="option"
    >
      <div
        class="content"
      >
        <div
          data-testid="tf-true-false-option-1-textfield"
        >
          <input
            data-testid="check-true-false-option-1-textfield"
            type="checkbox"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TrueFalseOptions > renders and matches snapshot 1`] = `
<div>
  <div
    class="optionsContainer"
  >
    <div
      class="option"
    >
      <div
        class="content"
      >
        <div
          data-testid="tf-true-false-option-0-textfield"
        >
          <input
            data-testid="check-true-false-option-0-textfield"
            type="checkbox"
          />
        </div>
      </div>
    </div>
    <div
      class="option"
    >
      <div
        class="content"
      >
        <div
          data-testid="tf-true-false-option-1-textfield"
        >
          <input
            data-testid="check-true-false-option-1-textfield"
            type="checkbox"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
